[Inductor][CPP] Support more than one LocalBuffer

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129121
* #126967

**Summary**
Support more than 1 Local Buffer in an outer loop fused node and also the case when multi global buffers sharing usage of same local buffer.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

leslie-fang-intel committed:
[Inductor][CPP] Support more than one LocalBuffer

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffer_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffer_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffer_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffer_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp_utils.py:
```diff
@@ -397,10 +397,12 @@ class LocalBufferContext:
         self.local_buffers: Dict[str, ir.Buffer] = {}
-        # Map Local Buffer name to Global Buffer
-        self.local_to_global: Dict[str, ir.Buffer] = {}
+        # map global buffer name to global buffer
+        self.global_buffers: Dict[str, ir.Buffer] = {}
```
This is tracked by `V.graph`. Do we still need it here?

leslie-fang-intel replied:
I guess we may need it, for the case of GEMM Template, the global buffer of GEMM out may not be kept in `V.graph`.

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node and also the case when multi global buffers sharing usage of same local buffer.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp_utils.py:
```diff
@@ -492,9 +494,9 @@ def localize_nodes(
-        local_buffers = list(self.local_buffers.values())
-        global_buffers = list(self.local_to_global.values())
-        assert len(global_buffers[0].get_size()) == len(local_buffers[0].get_size())
+        for global_buffer_name, global_buffer in self.global_buffers.items():
+            local_buffer = self.global_to_local[global_buffer_name]
+            assert len(global_buffer.get_size()) == len(local_buffer.get_size())
```
Perhaps we don't need this strict check on size equivalence? Local buffer can be smaller than the global one?

leslie-fang-intel replied:
Sure, remove this check.

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3857,8 +3856,8 @@ def get_call_ranges(node: BaseSchedulerNode):
 
-            LocalBuffer = namedtuple("LocalBuffer", ["local_buf", "global_buf"])
-            local_buffers: List[LocalBuffer] = []
+            BufferMapper = namedtuple("BufferMapper", ["local_buf", "global_buf"])
```
It is 1:n mapping from local buffer to global ones. Is it better to make it a dict mapping local buffer to a list of global buffers?

leslie-fang-intel replied:
Thanks for the suggestion. Added a list of buffers to record all the local_buffers and a dict mapping each local buffer name to a list of global buffers.

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node and also the case when multi global buffers sharing usage of same local buffer.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3919,7 +3934,7 @@ def get_call_ranges(node: BaseSchedulerNode):
                     metrics.CppOuterLoopFusedCount(
                         len(cpp_kernel_proxy_list),
-                        local_buffer_number=len(local_buffers),
+                        local_buffer_number=len(list(scope.local_buffers.values())),
```
?
```suggestion
                        local_buffer_number=len(scope.local_buffers),
```

leslie-fang-intel replied:
Thanks for the comment and changed.

peterbell10 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3888,23 +3887,39 @@ def get_call_ranges(node: BaseSchedulerNode):
+                                for local_buf in local_buffers
+                                if local_buffer_layout == local_buf.layout
+                            ]
+                            return shared_buffers[0] if shared_buffers else None
```
Wouldn't you need to do a liveness analysis to determine if the buffer can be reused?

leslie-fang-intel replied:
Thanks for the comment. Did you mean to check if the local buffer has been released before reusing it? 

If so, I think it may not be necessary. Since we merge the outer loops for scheduler nodes inside an `OuterLoopFusedSchedulerNode` and allocate the local buffers at the beginning of codegen loops for an  `OuterLoopFusedSchedulerNode` here 
https://github.com/pytorch/pytorch/blob/a1371e5fdca2dc5cf9e00946d5fbe9edb2d36528/torch/_inductor/codegen/cpp.py#L2049-L2051

The lifecycle of a local buffer is along with the  `OuterLoopFusedSchedulerNode`  and here we only reuse the local buffer for global buffers inside the same `OuterLoopFusedSchedulerNode`. So I think when we decide to reuse this local buffer here, it has not been released.

peterbell10 replied:
I mean we can have the value that's held by the buffer being needed later on in the computation. This reuse would overwrite the value. e.g. Say you have:
1. write A
2. write B
3. read A
4. read B

We cannot use the same buffer to store A and B without corrupting 3.

leslie-fang-intel replied:
Thanks for the comment. Added the extra check and corresponding UT for this case. Please kindly help to take a look again.

- We will recode all the scheduler nodes which has been visited during the loop to find local buffer in `visited_scheduler_nodes`
- When we check if we can reuse current local buffer for another global buffer, we will check
  - for global buffer in global buffers which has been replaced by this local buffer
    - for all the users of this global buffer has already be visited in `visited_scheduler_nodes`

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node and also the case when multi global buffers sharing usage of same local buffer.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node and also the case when multi global buffers sharing usage of same local buffer.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node and also the case when multi global buffers sharing usage of same local buffer.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node and also the case when multi global buffers sharing usage of same local buffer.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node and also the case when multi global buffers sharing usage of same local buffer.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3920,23 +3921,49 @@ def is_contiguous_index(index_expr, sizes):
+                                    ]
+                                    if global_buffer.name is not None
+                                ):
+                                    shared_buffers.append(local_buf)
+                            return shared_buffers[0] if shared_buffers else None
```
NIT: you can stop searching once you find a match
```suggestion
                                    return local_buf
                            return None
```

leslie-fang-intel replied:
Thanks for the comment and changed.

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node and also the case when multi global buffers sharing usage of same local buffer.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node and also the case when multi global buffers sharing usage of same local buffer.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node and also the case when multi global buffers sharing usage of same local buffer.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge

leslie-fang-intel commented:
@pytorchbot revert -m "Broken trunk and need rebase"

leslie-fang-intel commented:
@pytorchbot revert -m "Broken trunk and need rebase" -c ghfirst

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node and also the case when multi global buffers sharing usage of same local buffer.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node and also the case when multi global buffers sharing usage of same local buffer.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node and also the case when multi global buffers sharing usage of same local buffer.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node and also the case when multi global buffers sharing usage of same local buffer.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node and also the case when multi global buffers sharing usage of same local buffer.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Support more than one LocalBuffer"


**Summary**
Support more than 1 Local Buffer in an outer loop fused node and also the case when multi global buffers sharing usage of same local buffer.

**TestPlan**
```
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_two_local_buffers_in_outer_loop_fusion
python -u -m pytest -s -v inductor/test_cpu_repro.py -k test_share_local_buffers_in_outer_loop_fusion
```

**Next Step**

- [✓] Support more than one Local Buffer/Global Buffer

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge -f "unrelated CI failure"
